flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPool.java

```diff
@@ -266,104 +279,367 @@ public void disconnectResourceManager() {
+	 * @param slotSharingManager responsible for the slot sharing group for which to allocate the slot
+	 * @param resourceProfile specifying the requirements for the requested slot
+	 * @param locationPreferences containing preferred TaskExecutors on which to allocate the slot
+	 * @param allowQueuedScheduling true if queued scheduling (the returned task slot must not be completed yet) is allowed, otherwise false
+	 * @return A {@link SlotSharingManager.MultiTaskSlotLocality} which contains the allocated {@link SlotSharingManager.MultiTaskSlot}
+	 * 		and its locality wrt the given location preferences
+	 * @throws NoResourceAvailableException if no task slot could be allocated
+	 */
+	private SlotSharingManager.MultiTaskSlotLocality allocateMultiTaskSlot(
+			AbstractID groupId,
+			SlotSharingManager slotSharingManager,
+			ResourceProfile resourceProfile,
+			Collection<TaskManagerLocation> locationPreferences,
+			boolean allowQueuedScheduling) throws NoResourceAvailableException {
+
+		// check first whether we have a resolved root slot which we can use
+		SlotSharingManager.MultiTaskSlotLocality multiTaskSlotLocality = slotSharingManager.getResolvedRootSlot(
+			groupId,
+			locationPreferences);
+
+		if (multiTaskSlotLocality != null && multiTaskSlotLocality.getLocality() == Locality.LOCAL) {
+			return multiTaskSlotLocality;
+		}
+
+		final SlotRequestId allocatedSlotRequestId = new SlotRequestId();
+		final SlotRequestId multiTaskSlotRequestId = new SlotRequestId();
+
+		// check whether we have an allocated slot available which we can use to create a new multi task slot in
+		final SlotAndLocality slotAndLocality = pollAndAllocateSlot(allocatedSlotRequestId, resourceProfile, locationPreferences);
+
+		if (slotAndLocality != null && (slotAndLocality.getLocality() == Locality.LOCAL || multiTaskSlotLocality == null)) {
+
+			final AllocatedSlot allocatedSlot = slotAndLocality.getSlot();
+			final SlotSharingManager.MultiTaskSlot multiTaskSlot = slotSharingManager.createRootSlot(
+				multiTaskSlotRequestId,
+				CompletableFuture.completedFuture(slotAndLocality.getSlot()),
+				allocatedSlotRequestId);
+
+			if (allocatedSlot.tryAssignPayload(multiTaskSlot)) {
+				return SlotSharingManager.MultiTaskSlotLocality.of(multiTaskSlot, slotAndLocality.getLocality());
 			} else {
-				LOG.debug("There was no slot allocation with {} to be cancelled.", requestId);
+				multiTaskSlot.release(new FlinkException("Could not assign payload to allocated slot " +
+					allocatedSlot.getAllocationId() + '.'));
 			}
 		}
 
-		return CompletableFuture.completedFuture(Acknowledge.get());
-	}
+		if (multiTaskSlotLocality != null) {
+			// prefer slot sharing group slots over unused slots
+			if (slotAndLocality != null) {
+				releaseSlot(
+					allocatedSlotRequestId,
+					null,
+					new FlinkException("Locality constraint is not better fulfilled by allocated slot."));
+			}
+			return multiTaskSlotLocality;
```

Comment:
Yes because it could not find a slot which is `LOCAL`. The alternative would be to ask the ResourceManager but it will only give you a slot with an `UNKNOWN` locality because it is uncompleted.